scheduler/src/cook/compute_cluster.clj

```diff
@@ -156,3 +159,312 @@
+  (map-vals compute-cluster-config-ent->compute-cluster-config current-db-config-ents))
+
+(defn get-db-configs
+  "Get the current dynamic cluster configurations from the database"
+  [db]
+  (->> (get-db-config-ents db)
+       (map-vals compute-cluster-config-ent->compute-cluster-config)))
+
+;TODO: when all clusters have state, change the ComputeCluster protocol to add state operations
+(defn compute-cluster->compute-cluster-config
+  "Calculate dynamic cluster configuration from a compute cluster"
+  [{:keys [state-atom state-locked?-atom name]
+    {{:keys [template base-path ca-cert]} :config} :cluster-definition}]
+  {:name name
+   :template template
+   :base-path base-path
+   :ca-cert ca-cert
+   :state @state-atom
+   :state-locked? @state-locked?-atom})
+
+;TODO: in the future all clusters will be dynamic and we shouldn't need this
+(defn get-dynamic-clusters
+  "Get the current in-memory dynamic clusters"
+  []
+  (->> @cluster-name->compute-cluster-atom
+       (filter (fn [[_ cluster]] (:dynamic-cluster-config? cluster)))
+       (into {})))
+
+(defn get-in-mem-configs
+  "Get the current in-memory dynamic cluster configurations"
+  []
+  (->> (get-dynamic-clusters)
+       (map-vals compute-cluster->compute-cluster-config)))
+
+(defn compute-current-configs
+  "Synthesize the current view of cluster configurations by looking at the current configurations in the database
+  and the current configurations in memory. Alert on any inconsistencies. Database wins on inconsistencies."
+  [current-db-configs current-in-mem-configs]
+  (let [[only-db-keys only-in-mem-keys both-keys] (util/diff-map-keys current-db-configs current-in-mem-configs)]
+    (doseq [only-db-key only-db-keys]
+      (when (not= :deleted (-> only-db-key current-db-configs :state))
+        (log/info "Database cluster configuration is missing from in-memory configs. Cluster is only in the database and is not deleted."
+                  {:cluster-name only-db-key :cluster (current-db-configs only-db-key)})))
+    (doseq [only-in-mem-key only-in-mem-keys]
+      (when (not= :deleted (-> only-in-mem-key current-in-mem-configs :state))
+        (log/error "In-memory cluster configuration is missing from the database. Cluster is only in memory and is not deleted."
+                   {:cluster-name only-in-mem-key :cluster (current-in-mem-configs only-in-mem-key)})))
+    (doseq [key both-keys]
+      (let [keys-to-keep-synced [:base-path :ca-cert :state]]
+        (when (not= (-> key current-db-configs (select-keys keys-to-keep-synced))
+                    (-> key current-in-mem-configs (select-keys keys-to-keep-synced)))
+          (log/error "Base path, CA cert, or state differ between in-memory and database cluster configurations."
+                     {:cluster-name key
+                      :in-memory-cluster (current-in-mem-configs key)
+                      :db-cluster (current-db-configs key)})))))
+  (merge current-in-mem-configs current-db-configs))
```

Comment:
*hmms* Does this mean the memory state gets overwritten by the db state? That's safe if we make sure to always update the DB state first, then update the in-memory state (and we should put a comment documenting that invariant later when we actually do that.)

Reply:
the db state wins, but we attempt to update the in memory state first
but it doesn't really matter which one we do. we still won't attempt to make any changes until the next update cycle or next REST call

Reply:
If you make a change (e.g., via REST), it updates the memory state, but doesn't get to the DB for some reason you will get back an error and support person will be alerted. So I think it's fine

Reply:
Excellent: Can you show me the code how/where that happens?